pkg/storer/fix_refcnt.go

```diff
@@ -0,0 +1,283 @@
+			reserveSeen++
+			_, exists := items[bri.Address.String()]
+			if exists {
+				items[bri.Address.String()].reserveCnt++
+				if items[bri.Address.String()].reserveCnt > reserveMax {
+					reserveMax = items[bri.Address.String()].reserveCnt
+				}
+				items[bri.Address.String()].batches = append(items[bri.Address.String()].batches, bri.BatchID)
+				reserveCnt++
+			}
+			if ticked {
+				ticked = false
+				logger.Info("..still scanning reserve", "scanned", reserveSeen, "reserved", reserveCnt, "maxDupes", reserveMax, "elapsed", time.Since(startReserve).Round(time.Second))
+			}
+			return false, nil
+		})
+		ticker.Stop()
+		logger.Info("scanned reserve", "reserved", reserveCnt, "total", reserveSeen, "elapsed", time.Since(startReserve).Round(time.Second))
+
+		logger.Info("Scanning Uploads")
+		uploadCnt := 0
+		uploadSeen := 0
+		startUpload := time.Now()
+		ticked = false
+		ticker.Reset(statusInterval)
+		_ = upload.IterateAllPushItems(store, func(address swarm.Address) (bool, error) {
+			uploadSeen++
+			_, exists := items[address.String()]
+			if exists {
+				items[address.String()].uploadCnt++
+				uploadCnt++
+			}
+			if ticked {
+				ticked = false
+				logger.Info("..still scanning uploads", "scanned", uploadSeen, "uploads", uploadCnt, "elapsed", time.Since(startUpload).Round(time.Second))
+			}
+			return false, nil
+		})
+		ticker.Stop()
+		logger.Debug("scanned uploads", "uploads", uploadCnt, "total", uploadSeen, "elapsed", time.Since(startUpload).Round(time.Second))
+
+		logger.Info("Scanning Pins")
+		pinScanCount := 0
+		pinCnt := 0
+		var pinMax uint32
+		startPins := time.Now()
+		ticked = false
+		ticker.Reset(statusInterval)
+		_ = pinstore.IteratePinnedChunks(store, func(addr swarm.Address) (bool, error) {
+			pinScanCount++
+			_, exists := items[addr.String()]
+			if exists {
+				items[addr.String()].pinCnt++
+				if items[addr.String()].pinCnt > pinMax {
+					pinMax = items[addr.String()].pinCnt
+				}
+				pinCnt++
+			} else if len(items) == total { // Don't complain about missing pins if we can't handle all chunks at once
+				logger.Debug("missing pinned chunk", "address", addr)
+			}
+			if ticked {
+				ticked = false
+				logger.Info("..still scanning pins", "scanned", pinScanCount, "pinned", pinCnt, "total", pinTotal, "maxDupes", pinMax, "elapsed", time.Since(startPins).Round(time.Second))
+			}
+			return false, nil
+		})
+		ticker.Stop()
+		if pinTotal == -1 {
+			pinTotal = pinScanCount
+		}
+		logger.Info("scanned pins", "pinned", pinCnt, "total", pinTotal, "elapsed", time.Since(startPins).Round(time.Second))
+
+		batch, err := store.Batch(ctx)
+		if err != nil {
+			logger.Error(err, "Failed to create batch")
+		}
+		batchHits := 0
+
+		logger.Info("Summarizing refCnts")
+		refs := make(map[uint32]uint32)
+		for _, item := range items {
+			refs[item.item.RefCnt]++
+			newRefCnt := item.pinCnt*100 + item.reserveCnt + item.cacheCnt + item.uploadCnt
```

Comment:
@ldeffenb  what is the reason for this multiplication? `item.pinCnt*100`

Reply:
That is one of my hacks that you don't need.  In my case, if a chunk is pinned, I want it to STAY in the sharky regardless of the bugs that might inappropriately decrement the refCnt by 1 here or there.   You'll just want pinCnt, and drop the *100.